### PR TITLE
Add Github Actions for MacOS

### PR DESCRIPTION
**Pull request by o01eg**

### CODE REVIEW

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      CMAKE_BUILD_PARALLEL_LEVEL: 2
+      HOMEBREW_NO_AUTO_UPDATE: 1
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v3
+      - name: Configure
+        run: |
+          mkdir build
+          cd build
+          cmake -G "Xcode" -D BOOST_PARALLEL_JOBS=$(sysctl hw.ncpu | awk '{print $2}') ..
```
Comment (Vezzra) [inline .github/workflows/macos.yml:24]: On my system, I only run `cmake ..` for the configuration step.

File: .github/workflows/macos.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      - name: Build
+        run: |
+          cd build
+          cmake --build . --config RelWithDebInfo -- -parallelizeTargets
```
Comment (Vezzra) [inline .github/workflows/macos.yml:28]: On my system, I only run `cmake --build . --config RelWithDebInfo`, without the `-- -parallelizeTargets` part.

**Review by Vezzra — COMMENTED** [review]
I use slightly different commands to build the SDK on my Mac, see my comments. Don't know if that is of any significance though.

**Comment by o01eg** [comment]
libpng fails with:

```
2022-05-20T09:37:01.9474010Z CMake Error in CMakeLists.txt:
2022-05-20T09:37:01.9474220Z   The custom command generating
2022-05-20T09:37:01.9474350Z 
2022-05-20T09:37:01.9474660Z     /Users/runner/work/freeorion-sdk/freeorion-sdk/build/libpng-prefix/src/libpng-build/pnglibconf.c
2022-05-20T09:37:01.9474880Z 
2022-05-20T09:37:01.9474980Z   is attached to multiple targets:
2022-05-20T09:37:01.9475110Z 
2022-05-20T09:37:01.9475170Z     genvers
2022-05-20T09:37:01.9475330Z     gensym
2022-05-20T09:37:01.9475480Z     genfiles
2022-05-20T09:37:01.9475580Z 
2022-05-20T09:37:01.9475710Z   but none of these is a common dependency of the other(s).  This is not
2022-05-20T09:37:01.9475980Z   allowed by the Xcode "new build system".
```

File: .github/workflows/macos.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      CMAKE_BUILD_PARALLEL_LEVEL: 2
+      HOMEBREW_NO_AUTO_UPDATE: 1
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v3
+      - name: Configure
+        run: |
+          mkdir build
+          cd build
+          cmake -G "Xcode" ..
```
Comment (Vezzra) [inline .github/workflows/macos.yml:24]: The CI build still fails, so you might try removing the `-G "Xcode` part too. As I said, I only do `cmake ..`without specifying any options, and that works perfectly fine on my system.

Judging by the error message of the failed CI build, it has something to do with a new Xcode build system:
```
CMake Error in CMakeLists.txt:
  The custom command generating

    /Users/runner/work/freeorion-sdk/freeorion-sdk/build/libpng-prefix/src/libpng-build/pnglibconf.c

  is attached to multiple targets:

    genvers
    gensym
    genfiles

  but none of these is a common dependency of the other(s).  This is not
  allowed by the Xcode "new build system".
```
Maybe not specifying Xcode as the generator system to use circumvents that problem?

**Comment by Vezzra** [comment]
> libpng fails with:
> 
> ```
> 2022-05-20T09:37:01.9474010Z CMake Error in CMakeLists.txt:
> 2022-05-20T09:37:01.9474220Z   The custom command generating
> 2022-05-20T09:37:01.9474350Z 
> 2022-05-20T09:37:01.9474660Z     /Users/runner/work/freeorion-sdk/freeorion-sdk/build/libpng-prefix/src/libpng-build/pnglibconf.c
> 2022-05-20T09:37:01.9474880Z 
> 2022-05-20T09:37:01.9474980Z   is attached to multiple targets:
> 2022-05-20T09:37:01.9475110Z 
> 2022-05-20T09:37:01.9475170Z     genvers
> 2022-05-20T09:37:01.9475330Z     gensym
> 2022-05-20T09:37:01.9475480Z     genfiles
> 2022-05-20T09:37:01.9475580Z 
> 2022-05-20T09:37:01.9475710Z   but none of these is a common dependency of the other(s).  This is not
> 2022-05-20T09:37:01.9475980Z   allowed by the Xcode "new build system".
> ```

Sorry, didn't see that reply of yours before I made my comment above - but yes, that seems to be the problem. The build scripts for libpng seem to set up some kind of dependencies which this new Xcode build systems does not allow. So maybe not specifiying Xcode as generator solves that problem.

Can't say for sure though. I'm still on macOS 10.14, not 10.15, so the issue might have to do with the newer macOS system and the toolchain that ships with it, and nothing to do with wether you specify Xcode as generator or not...

**Comment by Vezzra** [comment]
@o01eg, seems to work now.

**Comment by o01eg** [comment]
@Vezzra Could you also test generated SDK? https://github.com/freeorion/freeorion-sdk/actions/runs/2357550428

**Comment by Vezzra** [comment]
@o01eg,

> Could you also test generated SDK?

Tried to build master with the generated SDK, which failed. Build log attached:

[build.log.zip](https://github.com/freeorion/freeorion-sdk/files/8740354/build.log.zip)

**Comment by o01eg** [comment]
Not sure what was broken here: 

```
CompileC /Users/user/SoftwareProjekte/FO/main/build/FreeOrion.build/Release/freeorionparse.build/Objects-normal/x86_64/xcode_dummy.o /Users/user/SoftwareProjekte/FO/main/build/xcode_dummy.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler
    cd /Users/user/SoftwareProjekte/FO/main/src
    export LANG=en_US.US-ASCII
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -x c++ -target x86_64-apple-macos10.12 -fmessage-length=0 -fdiagnostics-show-note-include-stack -fmacro-backtrace-limit=0 -Wno-trigraphs -fpascal-strings -O3 -Wno-missing-field-initializers -Wno-missing-prototypes -Wno-return-type -Wno-non-virtual-dtor -Wno-overloaded-virtual -Wno-exit-time-destructors -Wno-missing-braces -Wparentheses -Wswitch -Wno-unused-function -Wno-unused-label -Wno-unused-parameter -Wno-unused-variable -Wunused-value -Wno-empty-body -Wno-uninitialized -Wno-unknown-pragmas -Wno-shadow -Wno-four-char-constants -Wno-conversion -Wno-constant-conversion -Wno-int-conversion -Wno-bool-conversion -Wno-enum-conversion -Wno-float-conversion -Wno-non-literal-null-conversion -Wno-objc-literal-conversion -Wno-shorten-64-to-32 -Wno-newline-eof -Wno-c++11-extensions -DCMAKE_INTDIR=\"Release\" -Dfreeorionparse_EXPORTS -DFREEORION_MACOSX -DBOOST_DATE_TIME_NO_LOCALE -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -fasm-blocks -fstrict-aliasing -Wdeprecated-declarations -Winvalid-offsetof -Wno-sign-conversion -Wno-infinite-recursion -Wno-move -Wno-comma -Wno-block-capture-autoreleasing -Wno-strict-prototypes -Wno-range-loop-analysis -Wno-semicolon-before-method-body -I/Users/user/SoftwareProjekte/FO/main/build/Release/include -I/Users/user/SoftwareProjekte/FO/main/build/FreeOrion.build/Release/freeorionparse.build/DerivedSources-normal/x86_64 -I/Users/user/SoftwareProjekte/FO/main/build/FreeOrion.build/Release/freeorionparse.build/DerivedSources/x86_64 -I/Users/user/SoftwareProjekte/FO/main/build/FreeOrion.build/Release/freeorionparse.build/DerivedSources -Wmost -Wno-four-char-constants -Wno-unknown-pragmas -F/Users/user/SoftwareProjekte/FO/main/build/Release -DNDEBUG -fPIC -fvisibility=hidden -std=c++1z -MMD -MT dependencies -MF /Users/user/SoftwareProjekte/FO/main/build/FreeOrion.build/Release/freeorionparse.build/Objects-normal/x86_64/xcode_dummy.d --serialize-diagnostics /Users/user/SoftwareProjekte/FO/main/build/FreeOrion.build/Release/freeorionparse.build/Objects-normal/x86_64/xcode_dummy.dia -c /Users/user/SoftwareProjekte/FO/main/build/xcode_dummy.cpp -o /Users/user/SoftwareProjekte/FO/main/build/FreeOrion.build/Release/freeorionparse.build/Objects-normal/x86_64/xcode_dummy.o
ld: warning: directory not found for option '-L/Users/user/SoftwareProjekte/FO/main/build/Release'
ld: warning: directory not found for option '-F/Users/user/SoftwareProjekte/FO/main/build/Release'
ld: can't open output file for writing: /Users/user/SoftwareProjekte/FO/main/build/Release/libfreeorionparse.dylib, errno=2 for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by o01eg** [comment]
Not sure but maybe it because CI builds SDK on macos-10.15 and you use macos-10.15.

**Comment by Vezzra** [comment]
I have to admit, I'm stumped - these errors don't make sense. The `/Users/user/SoftwareProjekte/FO/main/build/Release` folder definitely gets created properly, and other build artefacts are placed there as expected (`libfreeorioncommon.dylib`, `libGiGi.dylib` etc.).

So why the warnings that the directory can't be found for the `-L` and `-F` options and why `libfreeorionparse.dylib` can't be created? I have no idea what's going on here...

**Comment by Vezzra** [comment]
> Not sure but maybe it because CI builds SDK on macos-10.15 and you use macos-10.15.

Well, that's possible, but would be rather strange - the SDK should not be usable only for macOS versions equal or greater than the one it was built on. That would be quite a problem...

Since when do we use the macOS 10.15 image for the CI? Hasn't the latest release version of the SDK already been build on 10.15? That version works on my system...

**Comment by o01eg** [comment]
Doesn't SDK for releases build on Travis CI where MacOS 10.14 still available?

Current Github Actions CI currently to test PRs.

**Comment by Vezzra** [comment]
> Doesn't SDK for releases build on Travis CI where MacOS 10.14 still available?

Hm, possible. Raises the question if the SDKs build on Travis on macOS 10.14 only work on macOS 10.14 and higher, which would be bad. IIRC, officially you should be able to build FO on macOS 10.9 and higher (or was it 10.12...?).

**Pull request merged**